Negative numbers on yend that make the library crash

Hey, there is a special case that is causing me an error using the library:

I see that when src Is equals to `yend size=750000 part=8525 pcrc32=ACBE5763\r\n.\r\n`
The values returned by decodedDecremental are 
`
nd: 0
ns: 1
end: 1
`

With these values, this part of the code: https://github.com/mnightingale/rapidyenc/blob/8792729266a137af0d5b5896b703e2e5607c4578/decoder.go#L191,
decrements the ns - 2 making it negative number and therefore the library crash with:

`
panic: runtime error: slice bounds out of range [-1:] goroutine 474 [running]: github.com/mnightingale/rapidyenc.(*Decoder).Transform(0xc0002465a0, {0xc002af3000, 0x1000, 0x1000}, {0xc000172000, 0x2c, 0x1000}, 0x0) /go/pkg/mod/github.com/mnightingale/rapidyenc@v0.0.0-20231127201145-8792729266a1/decoder.go:206 +0x915 github.com/mnightingale/rapidyenc.(*Decoder).Read(0xc0002465a0, {0xc0045651b0, 0x20e50, 0xc0002465a0?}) /go/pkg/mod/github.com/mnightingale/rapidyenc@v0.0.0-20231127201145-8792729266a1/decoder.go:142 
`

I solved by adding a check after the switch case in which
`
if nSrc < 0 {
	return nDst, 1, io.EOF
}
`

But not sure if is correct or not because I don't fully understand the code. Do you think is a good solution? I can do a pr for it

--- a/decoder.go
+++ b/decoder.go
@@ -195,7 +195,12 @@ transform:
 			nSrc += ns - 3
 			d.body = false
 		default:
-			nSrc += ns
+			if d.State == StateCRLFEQ {
+				d.State = StateCRLF
+				nSrc += ns - 1
+			} else {
+				nSrc += ns
+			}
 			return nDst, nSrc, transform.ErrShortSrc
 		}
 	}

--- a/decoder_test.go
+++ b/decoder_test.go
@@ -8,6 +8,7 @@ import (
 	"github.com/stretchr/testify/require"
 	"hash/crc32"
 	"io"
+	"strings"
 	"testing"
 )
 
@@ -35,6 +36,47 @@ func TestDecode(t *testing.T) {
 
 			dec := AcquireDecoder()
 			dec.SetReader(encoded)
+			b := bytes.NewBuffer(nil)
+			n, err := io.Copy(b, dec)
+			require.Equal(t, int64(len(raw)), n)
+			require.NoError(t, err)
+			require.Equal(t, raw, b.Bytes())
+			require.Equal(t, tc.crc, dec.Meta().Hash)
+			require.Equal(t, int64(len(raw)), dec.Meta().End)
+			ReleaseDecoder(dec)
+		})
+	}
+}
+
+// TestSplitReads tests "=yend" split across reads
+func TestSplitReads(t *testing.T) {
+	cases := []struct {
+		name string
+		raw  string
+		crc  uint32
+	}{
+		{"foobar", "foobar", 0x9EF61F95},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			raw := []byte(tc.raw)
+
+			enc := NewEncoder()
+			encoded := enc.Encode(raw)
+
+			readers := make([]io.Reader, 0)
+			readers = append(readers, strings.NewReader(fmt.Sprintf("=ybegin part=%d line=128 size=%d name=%s\r\n", 1, len(raw), "foo")))
+			readers = append(readers, strings.NewReader(fmt.Sprintf("=ypart begin=%d end=%d\r\n", 1, len(raw)+1)))
+			readers = append(readers, strings.NewReader(string(encoded)))
+			readers = append(readers, strings.NewReader("\r\n="))
+			readers = append(readers, strings.NewReader(fmt.Sprintf("yend size=%d part=%d pcrc32=%08x\r\n", len(raw), 1, crc32.ChecksumIEEE(raw))))
+			readers = append(readers, strings.NewReader(".\r\n"))
+
+			reader := io.MultiReader(readers...)
+
+			dec := AcquireDecoder()
+			dec.SetReader(reader)
 			b := bytes.NewBuffer(nil)
 			n, err := io.Copy(b, dec)
 			require.Equal(t, int64(len(raw)), n)

--- a/decoder_test.go
+++ b/decoder_test.go
@@ -68,7 +68,7 @@ func TestSplitReads(t *testing.T) {
 			readers := make([]io.Reader, 0)
 			readers = append(readers, strings.NewReader(fmt.Sprintf("=ybegin part=%d line=128 size=%d name=%s\r\n", 1, len(raw), "foo")))
 			readers = append(readers, strings.NewReader(fmt.Sprintf("=ypart begin=%d end=%d\r\n", 1, len(raw)+1)))
-			readers = append(readers, strings.NewReader(string(encoded)))
+			readers = append(readers, bytes.NewReader(encoded))
 			readers = append(readers, strings.NewReader("\r\n="))
 			readers = append(readers, strings.NewReader(fmt.Sprintf("yend size=%d part=%d pcrc32=%08x\r\n", len(raw), 1, crc32.ChecksumIEEE(raw))))
 			readers = append(readers, strings.NewReader(".\r\n"))

